kernel/kernel-api/src/main/java/io/delta/kernel/Scan.java

```diff
@@ -0,0 +1,95 @@
+ */
+public interface Scan {
+    /**
+     * Get an iterator of data files to scan.
+     *
+     * @param tableClient {@link TableClient} instance to use in Delta Kernel.
+     * @return data in {@link ColumnarBatch} format. Each row correspond to one scan file.
+     */
+    CloseableIterator<ColumnarBatch> getScanFiles(TableClient tableClient);
+
+    /**
+     * Get the remaining filter the Delta Kernel can not guarantee the data returned by it
+     * satisfies the filter. This filter is used by Delta Kernel to do data skipping whenever
+     * possible.
+     * @return Remaining filter as {@link Expression}.
+     */
+    Expression getRemainingFilter();
+
+    /**
+     * Get the scan state associate with the current scan. This state is common to all survived
+     * files.
+     *
+     * @param tableClient {@link TableClient} instance to use in Delta Kernel.
+     * @return Scan state in {@link Row} format.
+     */
+    Row getScanState(TableClient tableClient);
+
+    /**
+     * Get the data from a given scan files with using the connector provider {@link TableClient}.
+     *
+     * @param tableClient Connector provided {@link TableClient} implementation.
+     * @param scanState Scan state returned by {@link Scan#getScanState(TableClient)}
+     * @param scanFileRowIter an iterator of {@link Row}s. Each {@link Row} represents one scan file
+     *                        from the {@link ColumnarBatch} returned by
+     *                        {@link Scan#getScanFiles(TableClient)}
+     * @param filter An optional filter that can be used for data skipping while reading the
+     *               scan files.
+     * @return Data read from the scan file as an iterator of {@link ColumnarBatch}es and
+     *         {@link ColumnVector} pairs. The {@link ColumnVector} represents a selection vector
+     *         of type boolean that has the same size as the data {@link ColumnarBatch}. A value
+     *         of true at a given index indicates the row with the same index from the data
+     *         {@link ColumnarBatch} should be considered and a value of false at a given index
+     *         indicates the row with the same index in data {@link ColumnarBatch} should be
+     *         ignored. It is the responsibility of the caller to close the iterator.
+     *
+     * @throws IOException when error occurs reading the data.
+     */
+    static CloseableIterator<DataReadResult> readData(
+            TableClient tableClient,
+            Row scanState,
+            CloseableIterator<Row> scanFileRowIter,
+            Optional<Expression> filter) throws IOException {
+
+        StructType readSchema = Utils.getPhysicalSchema(scanState);
+
+        ParquetHandler parquetHandler = tableClient.getParquetHandler();
+
+        CloseableIterator<FileReadContext> filesReadContextsIter =
+                parquetHandler.contextualizeFileReads(
+                        scanFileRowIter,
+                        Literal.TRUE);
```

Comment:
This should be `filter` (or `Literal.True` if is empty) I think?

Reply:
Yep, it should be filter.orElse(literal.TRUE).